statsmodels/tsa/statespace/regime_switching/kim_filter.py

```diff
@@ -0,0 +1,635 @@
+        obs_intercepts = self._prepare_data_for_regimes(obs_intercepts,
+                self.per_regime_dims['obs_intercepts'])
+        obs_covs = self._prepare_data_for_regimes(obs_covs,
+                self.per_regime_dims['obs_covs'])
+        transitions = self._prepare_data_for_regimes(transitions,
+                self.per_regime_dims['transitions'])
+        state_intercepts = self._prepare_data_for_regimes(state_intercepts,
+                self.per_regime_dims['state_intercepts'])
+        selections = self._prepare_data_for_regimes(selections,
+                self.per_regime_dims['selections'])
+        state_covs = self._prepare_data_for_regimes(state_covs,
+                self.per_regime_dims['state_covs'])
+
+        # Kalman filters for each regime
+
+        kwargs['alternate_timing'] = True
+        self._regime_kalman_filters = [KalmanFilter(k_endog, k_states,
+                dtype=dtype, loglikelihood_burn=loglikelihood_burn,
+                design=designs[i], obs_intercept=obs_intercepts[i],
+                obs_cov=obs_covs[i], transition=transitions[i],
+                state_intercept=state_intercepts[i], selection=selections[i],
+                state_cov=state_covs[i], **kwargs) for i in range(k_regimes)]
+
+        self.set_regime_switch_probs(regime_switch_probs)
+
+    def set_regime_switch_probs(self, regime_switch_probs):
+
+        dtype = self._dtype
+        k_regimes = self._k_regimes
+
+        if regime_switch_probs is not None:
+            regime_switch_probs = np.asarray(regime_switch_probs, dtype=dtype)
+            if regime_switch_probs.shape != (k_regimes, k_regimes):
+                raise ValueError('Regime switching matrix should have shape'
+                        ' (k_regimes, k_regimes)')
+            if not self._is_left_stochastic(regime_switch_probs):
+                raise ValueError(
+                        'Provided regime switching matrix is not stochastic')
+            self._regime_switch_probs = regime_switch_probs
+        else:
+            self._regime_switch_probs = np.identity(k_regimes, dtype=dtype)
+
+    def _prepare_data_for_regimes(self, data, per_regime_dims):
+
+        k_regimes = self._k_regimes
+
+        if data is None:
+            return [None for _ in range(k_regimes)]
+
+        data = np.asarray(data, dtype=self._dtype)
+
+        if len(data.shape) == per_regime_dims:
+            return [data for _ in range(k_regimes)]
+
+        if data.shape[0] != k_regimes:
+            raise ValueError('First dimension is not k_regimes')
+
+        return data
+
+    def __getitem__(self, key):
+
+        if key == 'regime_switch_probs':
+            return self._regime_switch_probs
+
+        if key not in self.per_regime_dims and \
+                key + 's' not in self.per_regime_dims:
+            raise IndexError('"%s" is an invalid state space matrix name.' \
+                    % key)
+
+        if key[-1] == 's':
```

Comment:
My feeling is that allowing the plural form might be more trouble than it's worth in terms of future maintenance, because it might mean we need to support plurals elsewhere and then we forget, etc. I like the idea, but I've reversed course on things like this before because of the headache it's caused.